tests/framework/scheduler.py

```diff
@@ -0,0 +1,329 @@
+may be restricted to running sequentially, if they are per se
+concurrency-unsafe. See `PytestScheduler.pytest_runtestloop()`.
+
+Scheduling is achieved by overriding the pytest run loop (i.e.
+`pytest_runtestloop()`), and splitting the test session item list across
+multiple `fork()`ed worker processes. Since no user code is run before
+`pytest_runtestloop()`, each worker becomes a pytest session itself.
+Reporting is disabled for worker process, each worker sending its results
+back to the main / server process, via an IPC pipe, for aggregation.
+"""
+
+import multiprocessing as mp
+import os
+import re
+import sys
+from random import random
+from select import select
+from time import sleep
+
+from . import mpsing  # pylint: disable=relative-beyond-top-level
+
+
+class PytestScheduler(mpsing.MultiprocessSingleton):
+    """A pretty custom test execution scheduler."""
+
+    def __init__(self):
+        """Initialize the scheduler.
+
+        Not to be called directly, since this is a singleton. Use
+        `PytestScheduler.instance()` to get the scheduler object.
+        """
+        super().__init__()
+        self._mpsings = [self]
+        self.session = None
+
+    def register_mpsing(self, mpsing_inst):
+        """Register a multi-process singleton object.
+
+        Since the scheduler will be handling the main testing loop, it needs
+        to be aware of any multi-process singletons that must be serviced
+        during the test run (i.e. polled and allowed to handle method
+        execution in the server context).
+        """
+        self._mpsings.append(mpsing_inst)
+
+    @staticmethod
+    def do_pytest_addoption(parser):
+        """Pytest hook. Add concurrency command line option."""
+        avail_cpus = len(os.sched_getaffinity(0))
+        # Defaulting to a third of the available (logical) CPUs sounds like a
+        # good enough plan.
+        default = max(1, int(avail_cpus / 3))
+        parser.addoption(
+            "--concurrency",
+            "--concurrency",
+            dest="concurrency",
+            action="store",
+            type=int,
+            default=default,
+            help="Concurrency level (max number of worker processes to spawn)."
+        )
+
+    def pytest_sessionstart(self, session):
+        """Pytest hook. Called at pytest session start.
+
+        This will execute in the server context (before the tests are
+        executed).
+        """
+        self.session = session
+
+    def pytest_runtest_logreport(self, report):
+        """Pytest hook. Called whenever a new test report is ready.
+
+        This will execute in the worker / child context.
+        """
+        self._add_report(report)
+
+    def pytest_runtestloop(self, session):
+        """Pytest hook. The main test scheduling and running loop.
+
+        Called in the server process context.
+        """
+        max_concurrency = self.session.config.option.concurrency
+        schedule = [
```

Comment:
nit: this dict could be loaded from  a config file

Reply:
True, but IMO, config files are only useful if there's only one of them. Or at least, all config parameters being centralized somewhere. For this project though, configuration is done inline, with the various parameters being defined at the point of usage (i.e. in the same code file that uses them).